tests/modcc/test_parser.cpp

```diff
@@ -15,481 +79,413 @@ TEST(Parser, full_file) {
+    const char* good_expr[] = {
+        "((celsius-22)/10)      ",
+        "((celsius-22)+10)      ",
+        "(x+2)                  ",
+        "((x))                  ",
+        "(((x)))                ",
+        "(x + (x * (y*(2)) + 4))",
+    };
 
-        EXPECT_EQ(e, nullptr);
-        EXPECT_EQ(p.status(), lexerStatus::error);
+    for (auto& text: good_expr) {
+        EXPECT_TRUE(check_parse(&Parser::parse_parenthesis_expression, text));
+    }
 
-        #ifdef VERBOSE_TEST
-        if(e) std::cout << e->to_string() << std::endl;
-        if(p.status()==lexerStatus::error)
-            std::cout << "in " << cyan(bad_expression) << "\t" << p.error_message() << std::endl;
-        #endif
+    const char* bad_expr[] = {
+        "(x             ",
+        "((x+3)         ",
+        "(x+ +)         ",
+        "(x=3)          ",  // assignment inside parenthesis isn't allowed
+        "(a + (b*2^(x)) ",  // missing closing parenthesis
+    };
+
+    for (auto& text: bad_expr) {
+        EXPECT_TRUE(check_parse_fail(&Parser::parse_parenthesis_expression, text));
     }
+}
 
-    {
-        Parser p("LOCAL x, ");
-        auto e = p.parse_local();
+// test parsing of line expressions
+TEST(Parser, parse_line_expression) {
+    const char* good_expr[] = {
+        "qt=q10^((celsius-22)/10)"
+        "x=2        ",
+        "x=2        ",
+        "x = -y\n   "
+        "x=2*y      ",
+        "x=y + 2 * z",
+        "x=(y + 2) * z      ",
+        "x=(y + 2) * z ^ 3  ",
+        "x=(y + 2 * z ^ 3)  ",
+        "foo(x+3, y, bar(21.4))",
+        "y=exp(x+3) + log(exp(x/y))",
+        "a=x^y^z",
+        "a=x/y/z"
+    };
 
-        EXPECT_EQ(e, nullptr);
-        EXPECT_EQ(p.status(), lexerStatus::error);
+    for (auto& text: good_expr) {
+        EXPECT_TRUE(check_parse(&Parser::parse_line_expression, text));
+    }
 
-        #ifdef VERBOSE_TEST
-        if(e) std::cout << e->to_string() << std::endl;
-        if(p.status()==lexerStatus::error)
-            std::cout << "in " << cyan(bad_expression) << "\t" << p.error_message() << std::endl;
-        #endif
+    const char* bad_expr[] = {
+        "x=2+        ",      // incomplete binary expression on rhs
+        "x=          ",      // missing rhs of assignment
+        "x=)y + 2 * z",
+        "x=(y + 2    ",
+        "x=(y ++ z   ",
+        "x/=3        ",      // compound binary expressions not supported
+        "foo+8       ",      // missing assignment
+        "foo()=8     ",      // lhs of assingment must be an lvalue
+    };
+
+    for (auto& text: bad_expr) {
+        EXPECT_TRUE(check_parse_fail(&Parser::parse_line_expression, text));
     }
 }
 
-TEST(Parser, parse_unary_expression) {
-    std::vector<const char*> good_expressions =
-    {
-"+x             ",
-"-x             ",
-"(x + -y)       ",
-"-(x - + -y)    ",
-"exp(x + y)     ",
-"-exp(x + -y)   ",
+TEST(Parser, parse_stoich_term) {
+    const char* good_pos_expr[] = {
+        "B", "B3", "3B3", "0A", "12A"
```

Comment:
what about `2e4` or `1E3` for example? These are ambiguous, if we have a variable named `e4` or `E3` respectively. From looking at the lexer and parser code, I think these might be a problem.